evm/vm/forks/frontier/vm_state.py

```diff
@@ -0,0 +1,320 @@
+    with vm_state.state_db() as state_db:
+        for account, beneficiary in computation.get_accounts_for_deletion():
+            # TODO: need to figure out how we prevent multiple selfdestructs from
+            # the same account and if this is the right place to put this.
+            vm_state.logger.debug('DELETING ACCOUNT: %s', encode_hex(account))
+
+            # TODO: this balance setting is likely superflous and can be
+            # removed since `delete_account` does this.
+            state_db.set_balance(account, 0)
+            state_db.delete_account(account)
+
+    return computation
+
+
+def _make_frontier_receipt(vm_state, transaction, computation):
+    # Reusable for other forks
+
+    logs = [
+        Log(address, topics, data)
+        for address, topics, data
+        in computation.get_log_entries()
+    ]
+
+    gas_remaining = computation.get_gas_remaining()
+    gas_refund = computation.get_gas_refund()
+    tx_gas_used = (
+        transaction.gas - gas_remaining
+    ) - min(
+        gas_refund,
+        (transaction.gas - gas_remaining) // 2,
+    )
+
+    gas_used = vm_state.block_header.gas_used + tx_gas_used
+
+    receipt = Receipt(
+        state_root=vm_state.block_header.state_root,
+        gas_used=gas_used,
+        logs=logs,
+    )
+
+    return receipt
+
+
+class FrontierVMState(BaseVMState):
+    computation_class = FrontierComputation
+
+    @staticmethod
+    def execute_transaction(vm_state, transaction):
+        computation = _execute_frontier_transaction(vm_state, transaction)
+        return computation, vm_state.block_header
+
+    @staticmethod
+    def make_receipt(vm_state, transaction, computation):
+        receipt = _make_frontier_receipt(vm_state, transaction, computation)
+        return receipt
+
+    def validate_block(self, block):
+        if not block.is_genesis:
+            parent_header = self.get_parent_header(block.header)
+
+            self._validate_gas_limit(block)
+            validate_length_lte(block.header.extra_data, 32, title="BlockHeader.extra_data")
+
+            # timestamp
+            if block.header.timestamp < parent_header.timestamp:
+                raise ValidationError(
+                    "`timestamp` is before the parent block's timestamp.\n"
+                    "- block  : {0}\n"
+                    "- parent : {1}. ".format(
+                        block.header.timestamp,
+                        parent_header.timestamp,
+                    )
+                )
+            elif block.header.timestamp == parent_header.timestamp:
+                raise ValidationError(
+                    "`timestamp` is equal to the parent block's timestamp\n"
+                    "- block : {0}\n"
+                    "- parent: {1}. ".format(
+                        block.header.timestamp,
+                        parent_header.timestamp,
+                    )
+                )
+
+        # XXX: Should these and some other checks be moved into
```

Comment:
Maybe open an issue to revisit this separately.

Reply:
>  XXX: Should these and some other checks be moved into VM.validate_block(), as they apply to all block flavours?

Ah, these two lines and this function were copied from `Block.validate()`. `VMState.validate_block()` became kind of `applying to all block flavors` now.
Another future issue is that this function would need to be refactored for sharding #256 since there's no `Collation.uncle`.